### PR TITLE
MGMT-4795 - operator should configure Assisted Service deployment to …

### PR DESCRIPTION
**Pull request by yevgeny-shnaidman**
…include mirror registries configuration

User will provide the configuration by creating a ConfigMap that will contain 2 entries:
ca-bundle.crt - contents of ceritificate bundle for mirror registries
registries.conf - contents of registries.conf file

He will update the AgentServiceConfig with the name of the ConfigMap
Operator will map the ConfigMap into the Assisted-Service

### CODE REVIEW

**Review by djzager — COMMENTED** [review]
Looks sane to me. One comment that really just helps users consuming the operator, and creating an AgentServiceConfig, through the UI.

File: internal/controller/api/v1beta1/agentserviceconfig_types.go
Comment (yevgeny-shnaidman) [reply internal/controller/api/v1beta1/agentserviceconfig_types.go:46]: fixed

File: internal/controller/api/v1beta1/agentserviceconfig_types.go
Comment (flaper87) [inline internal/controller/api/v1beta1/agentserviceconfig_types.go:48]: Do we want to use a `LocalObjectReference` here instead of a `string` ? 

I have seen a mix of usage between `Refs` and `Names` and it'd be nice to be more consistent. `ClusterDeployment` uses references for cases like this. Any opinions about it?

Here's an example: https://github.com/openshift/hive/blob/d5b066d39c7d00acb05769c4935c0f8ea3554210/apis/hive/v1/clusterdeployment_types.go#L89

File: internal/controller/controllers/agentserviceconfig_controller.go
Comment (flaper87) [inline internal/controller/controllers/agentserviceconfig_controller.go:320]:
```suggestion
	keysToValidate := []string{"ca-bundle.crt", MirrorRegistriesConfigFile}
```

Should `ca-bundle.crt` also be in a const?

File: internal/common/common.go
Comment (flaper87) [inline internal/common/common.go:21]: This one is only used in L23. Does it have to be a const? maybe there's a use case I haven't considered.

**Review by flaper87 — CHANGES_REQUESTED** [review]

**Review by djzager — CHANGES_REQUESTED** [review]
@mhrivnak pointed out recently that the `docs/operator.md` needed an overhaul (that part is out of scope for this PR), but I think it's a good idea for you to explain how to use this feature in "Deploying the operand".

File: internal/controller/controllers/agentserviceconfig_controller.go
Comment (yevgeny-shnaidman) [reply internal/controller/controllers/agentserviceconfig_controller.go:320]: in this case those are the keys of the configmap, and it is used only here. i prefer to leave them as is, since the rest of the code also does not use const for other keys of the deployment

File: internal/common/common.go
Comment (yevgeny-shnaidman) [reply internal/common/common.go:21]: since i need to use PATH and FILE, i would prefer if PATH would have been constructed also from const. In addition, in case we want to access that directory in the future, we will already have a const for it

File: internal/controller/api/v1beta1/agentserviceconfig_types.go
Comment (yevgeny-shnaidman) [reply internal/controller/api/v1beta1/agentserviceconfig_types.go:48]: it is a matter of convention, if we have one. i personally think that Name string is more clear for the user, but i can change it. @djzager @mhrivnak what do you think?

Comment (djzager) [reply internal/controller/api/v1beta1/agentserviceconfig_types.go:48]: If we are using `string` I think `MirrorRegistryConfigMapName` makes sense. If we use `LocalObjectReference` then I believe `MirrorRegistryConfigMapRef` is better. Since I've seen both `Name` and `Ref`, I don't believe that there is an established convention so as long as we are internally consistent I'm happy.

Comment (djzager) [reply internal/controller/api/v1beta1/agentserviceconfig_types.go:48]: I'm not certain that `LocalObjectReference` type buys us very much in this scenario. Is there something I'm missing @flaper87 ? Assuming that using `LocalObjectReference` doesn't add value, I say `string` is fine for us.

Comment (flaper87) [reply internal/controller/api/v1beta1/agentserviceconfig_types.go:48]: > I'm not certain that LocalObjectReference type buys us very much in this scenario. Is there something I'm missing @flaper87 ? Assuming that using LocalObjectReference doesn't add value, I say string is fine for us.

The only thing I'm after is consistency. Here we are using strings, but then in other places we use `LocalObjectReference`. https://github.com/openshift/assisted-service/blob/5ddcfc1625de87be709c08f42bc83c73753962ca/internal/controller/api/v1beta1/infraenv_types.go#L63

To be honest, I think using a `LocalObjectReference` is clearer, it's more explicit, and it sets all the expectations and requirements we need. 

> If we are using string I think MirrorRegistryConfigMapName makes sense. If we use LocalObjectReference then I believe MirrorRegistryConfigMapRef is better.

:+1: 

I'm happy to let this one go if others think it's not worth it but it seems to me like we are letting all these cases go in and missing the chance of being consistent. This, for the user, is way worse than the extra characters needed to write down the reference.

File: internal/common/common.go
Comment (flaper87) [reply internal/common/common.go:21]: > since i need to use PATH and FILE, i would prefer if PATH would have been constructed also from const. In addition, in case we want to access that directory in the future, we will already have a const for it

This sounds like an early optimization we don't need. and it makes reading the series of const variables more confusing. We can always split it and create the const in the future if needed. Not a big deal, I guess.

File: internal/controller/controllers/agentserviceconfig_controller.go
Comment (djzager) [reply internal/controller/controllers/agentserviceconfig_controller.go:320]: As I see it, either we use constants that we've defined or we don't specify the constants.

The controller as it is now needs to get better at defining important values as constants and documenting them so I wouldn't use the current state as an argument to not define constants :sunglasses: .

From my perspective there are two options:
1. Define all of the important mirror registry related constants in this controller (or `common.go`); document them and use them everywhere those values (mostly strings) would appear. Here that would mean defining a constant for `"ca-bundle.crt"` and using it here.
2. Drop the constants and just refer to them everywhere explicitly as `"ca-bundle.crt"` or `"/etc/pki/ca-trust/extracted/pem/mirror-ca.pem"`

File: internal/controller/api/v1beta1/agentserviceconfig_types.go
Comment (mhrivnak) [reply internal/controller/api/v1beta1/agentserviceconfig_types.go:48]: `LocalObjectReference` is probably the better option and matches the core API. Matching what hive does is also valuable. If we have other fields that are local `*Name`, I'd support changing them to `LocalObjectReference` for consistency.

Comment (mhrivnak) [reply internal/controller/api/v1beta1/agentserviceconfig_types.go:48]: To elaborate, the "better" part is AIUI the fact that context and usage is implicit in the type. We don't have to document for each field "This is the Name of a kube resource in the same namespace as this resource." It's implicit.

Comment (djzager) [reply internal/controller/api/v1beta1/agentserviceconfig_types.go:48]: Wasn't aware of other examples in this repo :fearful:. Since you put it this way this is my vote:

```suggestion
	MirrorRegistryConfigMapRef *corev1.LocalObjectReference `json:"mirrorRegistryConfigmapRef,omitempty"`
```

File: internal/common/common.go
Comment (yevgeny-shnaidman) [reply internal/common/common.go]: fixed

File: internal/controller/controllers/agentserviceconfig_controller.go
Comment (yevgeny-shnaidman) [reply internal/controller/controllers/agentserviceconfig_controller.go]: fixed

File: internal/controller/api/v1beta1/agentserviceconfig_types.go
Comment (yevgeny-shnaidman) [reply internal/controller/api/v1beta1/agentserviceconfig_types.go]: fixed

**Comment by yevgeny-shnaidman** [comment]
> @mhrivnak pointed out recently that the `docs/operator.md` needed an overhaul (that part is out of scope for this PR), but I think it's a good idea for you to explain how to use this feature in "Deploying the operand".

added a section regarding mirror registry to doc

File: docs/operator.md
Comment (djzager) [inline docs/operator.md:160]: it seems worth noting that it needs to be in the `assisted-installer` namespace. Trying to imagine the bug being written :sunglasses:

Comment (yevgeny-shnaidman) [reply docs/operator.md:160]: the notice regarding namespace is in line 164

File: internal/common/common.go
Comment (djzager) [inline internal/common/common.go:25]: https://golang.org/doc/effective_go#constants and https://golang.org/doc/effective_go#commentary make me think that we should be doing something like:

```suggestion
// important keys for handling a mirror registry
const (
	MirrorRegistriesCertificateFile = "mirror_ca.pem"
	MirrorRegistriesCertificatePath = "/etc/pki/ca-trust/extracted/pem/" + MirrorRegistriesCertificateFile
	MirrorRegistriesConfigDir = "/etc/containers"
	MirrorRegistriesConfigFile = "registries.conf"
	MirrorRegistriesConfigPath = MirrorRegistriesConfigDir + "/" + MirrorRegistriesConfigFile
)
```

File: internal/controller/controllers/common.go
Comment (djzager) [inline internal/controller/controllers/common.go:22]: https://golang.org/doc/effective_go#constants and https://golang.org/doc/effective_go#commentary make me think that we should be doing something like:

```suggestion
// important keys for handling a mirror registry
const (
	mirrorRegistryRefCertKey = "ca-bundle.crt"
	mirrorRegistryRefRegistryConfKey = "registries.conf"
	mirrorRegistryConfVolume = "mirror-registry-conf"
	mirrorRegistryCertVolume = "mirror-registry-ca"
)
```

**Review by djzager — COMMENTED** [review]
This all looks good to me. Few odds an ends left for me before approval.

File: internal/controller/controllers/agentserviceconfig_controller.go
Comment (yevgeny-shnaidman) [reply internal/controller/controllers/agentserviceconfig_controller.go:734]: fixed

File: internal/controller/controllers/agentserviceconfig_controller.go
Comment (yevgeny-shnaidman) [reply internal/controller/controllers/agentserviceconfig_controller.go:750]: fixed

File: internal/common/common.go
Comment (yevgeny-shnaidman) [reply internal/common/common.go:25]: fixed

File: docs/operator.md
Comment (yevgeny-shnaidman) [reply docs/operator.md]: fixed

File: docs/operator.md
Comment (yevgeny-shnaidman) [reply docs/operator.md]: fixed

**Review by flaper87 — APPROVED** [review]
lgtm

**Comment by osherdp** [comment]
/hold
to avoid spam

**Review by flaper87 — COMMENTED** [review]
/unhold
/approve

**Pull request merged**
